Add TLS support using haproxy

tipabu opened the pull request:
Have haproxy listen on both 443 and 80; client needs to be smart about whether it wants TLS or not.

Uses the PROXY protocol support introduced back in Swift 2.18.0; if you're testing an old version of Swift, you'll need to drop the `send-proxy` in `/etc/haproxy/haproxy.cfg`.

See also: #51

tipabu commented:
Potential follow-up work:

* Install the self-signed cert as a trusted cert with something like
  ```
  sudo mkdir -p /usr/local/share/ca-certificates/extra
  sudo cp /etc/ssl/private/saio.crt /usr/local/share/ca-certificates/extra/
  sudo update-ca-certificates
  ```
  so you don't have to add `-k` to your curl commands. Should also be able to drop `SWIFTCLIENT_INSECURE=true` from `~/.profile` and `insecure = yes` from `/etc/swift/test.conf`.
* Check for `/vagrant/ca.pem`; if found, use that to sign the saio cert (and make sure it's the cert getting installed above). This would let devs create and use a long-lived CA cert that they could install _outside_ of the VSAIO VM. This would also be more representative of real-world certificate usage.
* Use some X509v3 features when signing; in particular, Subject Alternative Name (which, as I recall, a lot of browsers want these days). I _think_ https://github.com/tipabu/saio-tls/blob/872090f/saio.conf#L15-L21 makes a good starting point?

clayg commented:
Probably need to clean out some of state with rebuildswift:

```
diff --git a/bin/rebuildswift b/bin/rebuildswift
index c0ccb74..d03ddbf 100755
--- a/bin/rebuildswift
+++ b/bin/rebuildswifte.
 
@@ -15,6 +15,7 @@
 # limitations under the Licens
 /vagrant/bin/cleanswift
-rm -fr /etc/swift/* /var/run/swift/* /var/cache/swift/* /etc/pykmip/* /etc/rsyncd.*
+rm -fr /etc/swift/* /var/run/swift/* /var/cache/swift/* /etc/pykmip/* \
+       /etc/rsyncd.* /etc/ssl/private/saio.* /etc/haproxy/haproxy.cfg
 sed 's/"full_reprovision": false/"full_reprovision": true/g' /etc/chef/dna.json > /tmp/reload.json
 sudo chef-solo -c /etc/chef/solo.rb -j /tmp/reload.json
```

Alternatively maybe something with the create guards & full_reprovision?

I really like the idea of trying to get the cert installed - if there's anything we could do to get rid of the SWIFTCLIENT_INSECURE (at least ON the saio) that'd probably be good.  I was able to get awscli working

```
[plugins]
endpoint = awscli_plugin_endpoint

[profile default]
aws_access_key_id = test:tester
aws_secret_access_key = testing
# signature_version can be either s3v4 (for v4 sigs) or s3 (for v2)
s3 =
	signature_version = s3v4
	endpoint_url = https://saio
s3api =
	signature_version = s3v4
	endpoint_url = https://saio
```

... but I always have to specify `--no-verify-ssl` on the console

tipabu commented:
I've never really used `rebuildswift` -- but OK, sure!

Got the cert installed so you don't need to do any sorts of no-verify stuff as long as you're in the VM. Had to [go digging](https://incognitjoe.github.io/adding-certs-to-requests.html) to sort out things for requests tho :-1:

FWIW, there are `ca_bundle` and `verify_ssl` options you can plumb into the aws config (either at the profile or s3/s3api level) for the host, and swiftclient supports `OS_CACERT` when you know the cert you want to trust.

I *did* notice a bit of a wart -- haproxy is still running even when your localrc *doesn't* have `SSL` enabled. Then since we've got the `send-proxy` flag, all requests 400 like `Bad request syntax ('PROXY TCP4 192.168.8.1 192.168.8.80 54842 443').` :disappointed:

Also, we might want to get on those follow-ups sooner rather than later; the missing SAN gets noisy:
```
vagrant@saio:~$ aws s3 mb s3://bucket 
/usr/lib/python3/dist-packages/urllib3/connection.py:344: SubjectAltNameWarning: Certificate for saio has no `subjectAltName`, falling back to check for a `commonName` for now. This feature is being removed by major browsers and deprecated by RFC 2818. (See https://github.com/shazow/urllib3/issues/497 for details.)
  SubjectAltNameWarning
make_bucket: bucket
vagrant@saio:~$ aws s3 cp swift/tox.ini s3://bucket
/usr/lib/python3/dist-packages/urllib3/connection.py:344: SubjectAltNameWarning: Certificate for saio has no `subjectAltName`, falling back to check for a `commonName` for now. This feature is being removed by major browsers and deprecated by RFC 2818. (See https://github.com/shazow/urllib3/issues/497 for details.)
  SubjectAltNameWarning
upload: swift/tox.ini to s3://bucket/tox.ini                   
vagrant@saio:~$ swift list
/usr/lib/python3/dist-packages/urllib3/connection.py:344: SubjectAltNameWarning: Certificate for saio has no `subjectAltName`, falling back to check for a `commonName` for now. This feature is being removed by major browsers and deprecated by RFC 2818. (See https://github.com/shazow/urllib3/issues/497 for details.)
  SubjectAltNameWarning
/usr/lib/python3/dist-packages/urllib3/connection.py:344: SubjectAltNameWarning: Certificate for saio has no `subjectAltName`, falling back to check for a `commonName` for now. This feature is being removed by major browsers and deprecated by RFC 2818. (See https://github.com/shazow/urllib3/issues/497 for details.)
  SubjectAltNameWarning
bucket
vagrant@saio:~$ swift list bucket
/usr/lib/python3/dist-packages/urllib3/connection.py:344: SubjectAltNameWarning: Certificate for saio has no `subjectAltName`, falling back to check for a `commonName` for now. This feature is being removed by major browsers and deprecated by RFC 2818. (See https://github.com/shazow/urllib3/issues/497 for details.)
  SubjectAltNameWarning
/usr/lib/python3/dist-packages/urllib3/connection.py:344: SubjectAltNameWarning: Certificate for saio has no `subjectAltName`, falling back to check for a `commonName` for now. This feature is being removed by major browsers and deprecated by RFC 2818. (See https://github.com/shazow/urllib3/issues/497 for details.)
  SubjectAltNameWarning
tox.ini
```
Not all func tests pass; we really ought to do something with https://review.opendev.org/#/c/571021/ and/or https://review.opendev.org/#/c/470427/ (in particular, I _think_ my trouble is in how we hardcode `is_secure=False` when getting our S3Connection).

tipabu commented on cookbooks/swift/recipes/setup.rb:
```diff
@@ -173,12 +178,14 @@
   action :create
 end
-file "/home/#{node['username']}/.aws/config" do
+template "/home/#{node['username']}/.aws/config" do
```
Oh right, I oughta do something similar for s3cmd's config above...

tipabu commented on cookbooks/swift/templates/default/etc/swift/proxy-server/proxy-server.conf.d/20_settings.conf.erb:
```diff
@@ -1,6 +1,9 @@
 bind_port = 8080
 
+[app:proxy-server]
+require_proxy_protocol = <%= @ssl %>
```
I'm less and less certain that this is a good idea -- at least, it doesn't seem to make a lot of sense for a dev environment. Can't run probe tests like this! Though I like the idea of using PROXY protocol when terminating TLS... I don't know that I can swing it for this :disappointed:

tipabu replied:
Got an upstream patch to resolve it -- I'm not so worried now.

tipabu commented:
OK:

* took out PROXY protocol so I can still run probe tests,
* added SAN support, and even
* switched to using an EC key for the latest hawtness.

Still have some problems, though:

* no PROXY protocol means I need to do some plumbing to get tempurl/staticweb/whoever-else to return https redirects;
* need to fix up s3cmd's config -- forgot about that -- but since we can now handle S3 requests on any of ports 80, 443, or 8080, it doesn't matter as much for now; and
* the s3 func tests are still problematic, but I think that'll have to be resolved upstream.

clayg commented:
I don't really understand why the proxy protocol is incompatible with probetests - I'm checking out the change to try and figure that out.

FWIW, I think it's reasonable for vsaio to support some configurations that we want to "test" that are incompatible with probe tests (e.g. we can do lots of disk configurations that cause skiptests in probe)

I'd rather be able to quickly get an production like ssl setup going for testing in dev with probetests failing (or skipping) than have a dev ssl setup that's not sufficiently like the ssl configs we care about. 

Perhaps SSL=true and SSL_PROXY_PROTO=true options?

tipabu commented:
> I don't really understand why the proxy protocol is incompatible with probetests

Yeah, tests pop something like
```
Traceback (most recent call last):
  File "/vagrant/swift/test/probe/common.py", line 390, in setUp
    proxy_ipport, self.ipport2server)
  File "/vagrant/swift/test/probe/common.py", line 121, in check_server
    ipport, 'test:tester', 'testing'))
  File "/vagrant/swift/test/probe/common.py", line 104, in _retry_timeout
    raise err
  File "/vagrant/swift/test/probe/common.py", line 97, in _retry_timeout
    return f(*args, **kwargs)
  File "/vagrant/swift/test/probe/common.py", line 85, in _check_proxy
    user, key)
  File "/vagrant/python-swiftclient/swiftclient/client.py", line 708, in get_auth
    timeout=timeout)
  File "/vagrant/python-swiftclient/swiftclient/client.py", line 556, in get_auth_1_0
    raise ClientException.from_response(resp, 'Auth GET failed', body)
swiftclient.exceptions.ClientException: Auth GET failed: http://127.0.0.1:8080/auth/v1.0 400 Invalid PROXY line 'GET /auth/v1.0 HTTP/1.1\r\n'
```
Following https://github.com/openstack/swift/commit/771963c92654338011d3dc907044b2ee3f405dd3 we go to the configured `bind_port`, but we hardcoded 8080 for a long time before that, at least since https://github.com/openstack/swift/commit/4a9d19197c5084c2a2a804ac2564950c913d5c2e

> Perhaps SSL=true and SSL_PROXY_PROTO=true options?

Not a bad plan -- I'll also see what it looks like to get some config into probe tests (upstream already has a `[probe_test]` section!) to route them through haproxy... and whether they still pass :thinking:

> Is there a tempauth change to fix the protocol returned as storage url? Is that what you were talking about with "redirects"?

Yeah, I've got some changes locally that'd address that; the meat of it is
```
diff --git a/cookbooks/swift/templates/default/etc/swift/proxy-server/proxy-server.conf.d/20_settings.conf.erb b/cookbooks/swift/templates/default/etc/swift/proxy-server/proxy-server.conf.d/20_settings.conf.erb
index 8bba279..bad4513 100644
--- a/cookbooks/swift/templates/default/etc/swift/proxy-server/proxy-server.conf.d/20_settings.conf.erb
+++ b/cookbooks/swift/templates/default/etc/swift/proxy-server/proxy-server.conf.d/20_settings.conf.erb
@@ -6,6 +6,7 @@ pipeline = catch_errors healthcheck proxy-logging domain_remap cache container_s
 
 [filter:tempauth]
 use = egg:swift#tempauth
+storage_url_scheme = <%= @ssl ? "https" : "http" %>
 user_admin_admin = admin .admin .reseller_admin
 user_test_tester = testing .admin
 user_test_tester2 = testing2 .admin
```

tipabu commented:
https://review.opendev.org/#/c/705579/ turned out not-terrible -- maybe we can keep PROXY protocol around...
